### PR TITLE
extension events data flow pipeline

### PR DESCRIPTION
**Pull request by nafees87n**

### CODE REVIEW

File: browser-extension/mv2/src/background/background.js [hunk 1]
```diff
@@ -974,6 +974,61 @@ BG.Methods.sendMessage = function (messageObject, callback) {
...
+  }
+
+  if (!BG.isAppOnline) {
+    return;
+  }
+
+  const extensionEventsMessage = {
+    action: RQ.EXTENSION_MESSAGES.EXTENSION_EVENTS,
+    events: {}, // it should fetched from a separate function only after all the checks and shouldn't be passed as params
+  };
```
Comment (nsrCodes) [inline browser-extension/mv2/src/background/background.js:1017]: We will not get an action from anywhere, will be triggered by a setInterval running in the background script itself

reference code around this commit -- https://github.com/requestly/requestly/pull/454/commits/651785841f82085b8a39d9be0392a583d63fb927

File: browser-extension/mv2/src/background/background.js [hunk 2]
```diff
@@ -974,6 +974,61 @@ BG.Methods.sendMessage = function (messageObject, callback) {
...
+      if (res.received) {
+        console.log("!!!debug", "ack in bg!! clear the events!!");
+      }
+    })
+    .catch(() => {
+      if (retryCounter < 3) {
+        BG.Methods.sendExtensionEvents(++retryCounter);
+      }
+    });
+};
```
Comment (nsrCodes) [inline browser-extension/mv2/src/background/background.js:1030]: unnecessary function, wasn't mentioned in initial architecture discussions

File: browser-extension/mv2/src/background/background.js [hunk 3]
```diff
@@ -974,6 +974,61 @@ BG.Methods.sendMessage = function (messageObject, callback) {
...
+
+  return Promise.race([
+    sendMessage(messageObject, tabId),
+    new Promise((_, reject) =>
+      setTimeout(() => reject(new Error("timeout")), timeout)
+    ),
+  ]);
+};
+
+BG.Methods.sendExtensionEvents = (tabId = null, retryCounter = 0) => {
```
Comment (nsrCodes) [inline browser-extension/mv2/src/background/background.js:998]: This function will not have any arguments for now

File: browser-extension/mv2/src/background/background.js [hunk 4]
```diff
@@ -974,6 +974,61 @@ BG.Methods.sendMessage = function (messageObject, callback) {
...
+  let appTabId = tabId;
+  if (!appTabId) {
+    chrome.tabs.query({ url: RQ.configs.WEB_URL + "/*" }, (tabs) => {
+      if (tabs.length > 0) {
+        appTabId = tabs[0].id;
+      } else {
+        BG.isAppOnline = false;
```
Comment (nsrCodes) [inline browser-extension/mv2/src/background/background.js:1005]: return after this and remove the extra if statement below

File: browser-extension/mv2/src/background/background.js [hunk 5]
```diff
@@ -974,6 +974,61 @@ BG.Methods.sendMessage = function (messageObject, callback) {
...
+    sendMessage(messageObject, tabId),
+    new Promise((_, reject) =>
+      setTimeout(() => reject(new Error("timeout")), timeout)
+    ),
+  ]);
+};
+
+BG.Methods.sendExtensionEvents = (tabId = null, retryCounter = 0) => {
+  let appTabId = tabId;
+  if (!appTabId) {
```
Comment (nsrCodes) [inline browser-extension/mv2/src/background/background.js:1000]: should be `if(BG.isAppOnline)`

File: browser-extension/mv2/src/background/background.js [hunk 6]
```diff
@@ -974,6 +974,61 @@ BG.Methods.sendMessage = function (messageObject, callback) {
   );
 };
 
+BG.Methods.sendMessageToApp = (messageObject, tabId, timeout = 2000) => {
+  const sendMessage = (messageObject, tabId) => {
```
Comment (nsrCodes) [inline browser-extension/mv2/src/background/background.js:978]: This function will not get the message object from anywhere

As discussed, this is not a reusable function for now and will also be responsible for reading events from local storage

so you still have to create a separate function called `sendEventsToTab` that does all this and also the batching of events to handle the edge cases / acknowledgement procedure of this function

File: browser-extension/mv2/src/background/background.js [hunk 7]
```diff
@@ -974,6 +974,61 @@ BG.Methods.sendMessage = function (messageObject, callback) {
...
+    return new Promise((resolve, reject) => {
+      chrome.tabs.sendMessage(tabId, messageObject, (response) => {
+        if (response) {
+          resolve(response);
+        } else {
+          reject(new Error("No response from app"));
```
Comment (nsrCodes) [inline browser-extension/mv2/src/background/background.js:984]: is this reject required?

**Review by nsrCodes — CHANGES_REQUESTED** [review]

File: browser-extension/mv2/src/background/background.js (hunk 6 above)
Comment (nafees87n) [reply browser-extension/mv2/src/background/background.js:978]: No, we anyway need a `sendMessageToApp` which will be a general, reusable function. Whatever event function we create, internally in the last step, it must call `sendMessageToApp` to send the event to App/UI.

File: browser-extension/mv2/src/background/background.js (hunk 7 above)
Comment (nsrCodes) [reply browser-extension/mv2/src/background/background.js:984]: remove reject, will be covered by timeout

File: browser-extension/mv2/src/modules/EventActions.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
+const EventActions = {};
+
+EventActions.sendExtensionEvents = async () => {
+  if (!BG.isAppOnline) {
+    return;
+  }
+
+  let ackReceived = false;
```
Comment (nsrCodes) [inline browser-extension/mv2/src/modules/EventActions.js:8]: no need to save this as a variable, 

we can just break the loop based on the response from `sendMessageToApp`

**Review by nsrCodes — CHANGES_REQUESTED** [review]

**Pull request merged**